### PR TITLE
Amesos : add MKL Cluster Sparse Solver interface

### PR DESCRIPTION
**Pull request by iyamazaki**
@trilinos/amesos 

## Motivation

This PR adds MKL Cluster Sparse Solver interface to Amesos

## Testing

Runnig ``ctest`` in ``amesos/test``:
```
    Start 1: Amesos_Test_Epetra_RowMatrix_MPI_4
1/8 Test #1: Amesos_Test_Epetra_RowMatrix_MPI_4 ...   Passed    9.86 sec
    Start 2: Amesos_Test_Epetra_CrsMatrix_MPI_4
2/8 Test #2: Amesos_Test_Epetra_CrsMatrix_MPI_4 ...   Passed    1.36 sec
    Start 3: Amesos_Test_MultipleSolves_MPI_4
3/8 Test #3: Amesos_Test_MultipleSolves_MPI_4 .....   Passed    1.78 sec
    Start 4: Amesos_Test_Detailed_MPI_4
4/8 Test #4: Amesos_Test_Detailed_MPI_4 ...........   Passed    1.85 sec
    Start 5: Amesos_Test_LAPACK_MPI_4
5/8 Test #5: Amesos_Test_LAPACK_MPI_4 .............   Passed    1.45 sec
    Start 6: Amesos_Test_KLU_MPI_4
6/8 Test #6: Amesos_Test_KLU_MPI_4 ................   Passed    1.64 sec
    Start 7: Amesos_Test_KLU_memleak_MPI_4
7/8 Test #7: Amesos_Test_KLU_memleak_MPI_4 ........   Passed    1.83 sec
    Start 8: Amesos_Test_Singular_MPI_4
8/8 Test #8: Amesos_Test_Singular_MPI_4 ...........   Passed    1.22 sec

100% tests passed, 0 tests failed out of 8
```
Running ``./amesos_test.exe CSS ${MATRIX_FILE} 1 1 0 0 1e+1 1e-1`` in ``Test_Basic``:
```
----------------------------------------------------------------------------
Amesos_CssMKL : Time to convert matrix to CssMKL format = 1.44169e-05 (s)
Amesos_CssMKL : Time to redistribute matrix = 0 (s)
Amesos_CssMKL : Time to redistribute vectors = 2.62167e-07 (s)
Amesos_CssMKL : Number of symbolic factorizations = 1
Amesos_CssMKL : Time for sym fact = 0.00354947 (s), avg = 0.00354947 (s)
Amesos_CssMKL : Number of numeric factorizations = 1
Amesos_CssMKL : Time for num fact = 0.000720826 (s), avg = 0.000720826 (s)
Amesos_CssMKL : Number of solve phases = 1
Amesos_CssMKL : Time for solve = 0.000457214 (s), avg = 0.000457214 (s)
----------------------------------------------------------------------------
==================================================================================================================

                                      TimeMonitor results over 8 processors

Timer Name                                MinOverProcs     MeanOverProcs    MaxOverProcs    MeanOverCallCounts    
------------------------------------------------------------------------------------------------------------------
Epetra_CrsMatrix::Multiply(TransA,x,y)    1.419e-05 (1)    2.131e-05 (1)    2.49e-05 (1)    2.131e-05 (1)         
TestSolver::Matrix Distribute             0.001101 (1)     0.00111 (1)      0.001115 (1)    0.00111 (1)           
TestSolver::Matrix Read                   0.05256 (1)      0.05605 (1)      0.05656 (1)     0.05605 (1)           
TestSolver::Numeric (CSS)                 0.2325 (2)       0.2325 (2)       0.2325 (2)      0.1162 (2)            
TestSolver::Solve   (CSS)                 0.03882 (2)      0.03883 (2)      0.03885 (2)     0.01942 (2)           
TestSolver::Symbolic(CSS)                 0.1731 (2)       0.1731 (2)       0.1731 (2)      0.08655 (2)           
Total                                     0 (1)            0 (1)            0 (1)           0 (1)                 
==================================================================================================================
```

## Additional Information
It uses TPL:
```
    -D TPL_ENABLE_CSS_MKL:BOOL=ON \
    -D   CSS_MKL_LIBRARY_DIRS="${MKLROOT}/lib/intel64" \
    -D   CSS_MKL_INCLUDE_DIRS="${MKLROOT}/include" \
```

### CODE REVIEW

File: packages/amesos/test/Test_Basic/Amesos_TestSolver.cpp [hunk 1]
```diff
@@ -225,7 +231,11 @@ int Amesos_TestSolver( Epetra_Comm &Comm, char *matrix_file,
     Comm.Barrier();
     
     A.Export(*serialA, exporter, Add);
-    assert(A.FillComplete()==0);    
+    //assert(A.FillComplete()==0);
+    if(A.FillComplete()!=0) {
+      exit(0);
+    }
+    delete serialA;
```
Comment (ndellingwood) [inline packages/amesos/test/Test_Basic/Amesos_TestSolver.cpp:238]: Does this need to be guarded by `if (distribute_matrix)` to balance the case below around L522? Why is this deleted here but conditionally deleted at L522, breaking the convention of the other items?

**Review by ndellingwood — COMMENTED** [review]
@iyamazaki I skimmed the code and have one question at a spot in the test. Otherwise, can you post manual test results with the new solver, and confirm that testing Pardiso without the new Pardiso CSS feature still passes? Thanks!

File: packages/amesos/test/Test_Basic/Amesos_TestSolver.cpp (hunk 1 above)
Comment (iyamazaki) [reply packages/amesos/test/Test_Basic/Amesos_TestSolver.cpp:238]: @ndellingwood. This is guarded by `if (distribued_matrix)' on L211? I am having challenge running this test with a large matrix, and was hoping freeing up memory will help.

**Comment by ndellingwood** [comment]
@trilinos/framework the PR has failed twice with the `PR_cuda` job, but looking at cdash results there are no errors posted, though it looks like the tests did not run. Can you help advise?

**Comment by iyamazaki** [comment]
@ndellingwood.. I am not sure how, but it now passed the tests..

File: packages/amesos/CMakeLists.txt [hunk 2]
```diff
@@ -35,6 +35,15 @@ IF(NOT TPL_ENABLE_MPI)
 ENDIF()
 ENDIF()
 
+# if using CSC, must be parallel and needs Pardiso_MKL
```
Comment (ndellingwood) [inline packages/amesos/CMakeLists.txt:38]: @iyamazaki is this set up so that `-DAmesos2_ENABLE_CSS_MKL` is automagically enabled only if `-DTPL_ENABLE_CSS_MKL=ON` is set? And can you confirm that if you enable MKL and Pardiso, but have `-DTPL_ENABLE_CSS_MKL=OFF`, that  `HAVE_AMESOS_CSS_MKL` is undefined/OFF?
I'm checking to prevent issues where an MKL version might not have CSS available (I'm thinking of the Albany issue)

Comment (iyamazaki) [reply packages/amesos/CMakeLists.txt:38]: @ndellingwood. We only require ``MPI``, and ``Pardiso_mkl`` is not needed. I also had a serial build with ``TPL_ENABLE_CSS_MKL``, and in that case, it will just turn off CSS.

**Comment by achauphan** [comment]
Looking into this now. It looks like something is turning Amesos off after being configured on, but only affects CUDA builds...?

**Comment by iyamazaki** [comment]
Hi, @achauphan. I do not know how, but our last PR tests all passed..

**Comment by sebrowne** [comment]
This was likely due to the autotester configuration transitions.  Regardless, I'm glad it passed.

Note that Amesos is disabled for the CUDA builds, as we deliberately disable any packages that will be deprecated next year for those builds (e.g. Amesos, Epetra).

**Review by ndellingwood — APPROVED** [review]
